protos/consensus.proto

```diff
@@ -0,0 +1,325 @@
+  ConsensusMessage message = 1;
+}
+
+message ConsensusOnMessageReceivedResponse {
+  enum Status {
+    STATUS_UNSET = 0;
+    OK = 1;
+    INTERNAL_ERROR = 2;
+  }
+
+  Status status = 1;
+}
+
+// Called when a new block is received and validated
+message ConsensusOnNewBlockReceivedRequest {
+  ConsensusBlock block = 1;
+}
+
+message ConsensusOnNewBlockReceivedResponse {
+  enum Status {
+    STATUS_UNSET = 0;
+    OK = 1;
+    INTERNAL_ERROR = 2;
+  }
+
+  Status status = 1;
+}
+
+// Called when a new peer is added
+message ConsensusOnAddPeerRequest {
+  string peer_id = 1;
+}
+
+message ConsensusOnAddPeerResponse {
+  enum Status {
+    STATUS_UNSET = 0;
+    OK = 1;
+    INTERNAL_ERROR = 2;
+  }
+
+  Status status = 1;
+}
+
+// Called when a peer is dropped
+message ConsensusOnDropPeerRequest {
+  string peer_id = 1;
+}
+
+message ConsensusOnDropPeerResponse {
+  enum Status {
+    STATUS_UNSET = 0;
+    OK = 1;
+    INTERNAL_ERROR = 2;
+  }
+
+  Status status = 1;
+}
+
+// Block Creation
+
+// Initialize a new block based on the current chain head and start
+// adding batches to it.
+message ConsensusInitializeBlockRequest {
+
+}
+
+message ConsensusInitializeBlockResponse {
+  enum Status {
+    STATUS_UNSET = 0;
+    OK = 1;
+    INTERNAL_ERROR = 2;
+    NOT_READY = 3;
+  }
+
+  Status status = 1;
+}
+
+// Stop adding batches to the current block and finalize it. Include
+// the given consensus data in the block. If this call is successful,
+// OnNewBlockReceived() will be called with the block.
+message ConsensusFinalizeBlockRequest {
+  bytes data = 1;
+}
+
+message ConsensusFinalizeBlockResponse {
+  enum Status {
+    STATUS_UNSET = 0;
+    OK = 1;
+    INTERNAL_ERROR = 2;
+    NOT_READY = 3;
+  }
+
+  Status status = 1;
+  string block_id = 2;
+}
+
+// Stop adding batches to the current block and abandon it.
+message ConsensusCancelBlockRequest {
+
+}
+
+message ConsensusCancelBlockResponse {
+  enum Status {
+    STATUS_UNSET = 0;
+    OK = 1;
+    INTERNAL_ERROR = 2;
+    NOT_READY = 3;
+  }
+
+  Status status = 1;
+}
+
+// Block Management
+
+// Set the chain head to the given block id.
+message ConsensusCommitBlockRequest {
+  string block_id = 1;
+}
+
+message ConsensusCommitBlockResponse {
+  enum Status {
+    STATUS_UNSET = 0;
+    OK = 1;
+    INTERNAL_ERROR = 2;
+    NOT_READY = 3;
+    NO_RESOURCE = 5;
+  }
+
+  Status status = 1;
+}
+
+// Remove the given block from the system. The main purpose of this
+// method is to allow the Consensus Engine to notify the rest of the
+// system that a block is invalid from the perspective of consensus.
+message ConsensusDropBlockRequest {
+  string block_id = 1;
+}
+
+message ConsensusDropBlockResponse {
+  enum Status {
+    STATUS_UNSET = 0;
+    OK = 1;
+    INTERNAL_ERROR = 2;
+    NOT_READY = 3;
+    NO_RESOURCE = 5;
+  }
+
+  Status status = 1;
+}
+
+// Queries
+
+// Read the current value of the setting.
+message ConsensusSettingGetRequest {
+  string setting = 1;
```

Comment:
Who handles what state root hash this is relative to?

Reply:
Probably need a head_id for that.